src/parsers/xml-processor/class-xmlprocessor.php

```diff
@@ -0,0 +1,4499 @@
+<?php
+
+namespace WordPress\XML;
+
+use WP_HTML_Span;
+use WP_HTML_Text_Replacement;
+
+use function WordPress\Encoding\utf8_codepoint_at;
+
+/**
+ * XML API: XMLProcessor class
+ *
+ * Scans through an XML document to find specific tags, then
+ * transforms those tags by adding, removing, or updating the
+ * values of the XML attributes within that tag (opener).
+ *
+ * It implements a subset of the XML 1.0 specification (https://www.w3.org/TR/xml/)
+ * and supports XML documents with the following characteristics:
+ *
+ * * XML 1.0
+ * * Well-formed
+ * * UTF-8 encoded
+ * * Not standalone (so can use external entities)
+ * * No DTD, DOCTYPE, ATTLIST, ENTITY, or conditional sections
+ *
+ * ### Possible future direction for this module
+ *
+ * The final goal is to support both 1.0 and 1.1 depending on the
+ * initial processing instruction (<?xml version="1.0" ?>). We're
+ * starting with 1.0, however, because most that's what most WXR
+ * files declare.
+ *
+ * @TODO: Include the cursor string in internal bookmarks and use it for seeking.
+ *
+ * @TODO: Track specific error states, expose informative messages, line
+ *        numbers, indexes, and other debugging info.
+ *
+ * @TODO: Skip over the following syntax elements:
+ *        * <!DOCTYPE, see https://www.w3.org/TR/xml/#sec-prolog-dtd
+ *        * <!ATTLIST, see https://www.w3.org/TR/xml/#attdecls
+ *        * <!ENTITY, see https://www.w3.org/TR/xml/#sec-entity-decl
+ *        * <!NOTATION, see https://www.w3.org/TR/xml/#sec-entity-decl
+ *        * Conditional sections, see https://www.w3.org/TR/xml/#sec-condition-sect
```

Comment:
skip over or bail? I guess if we ignore we can bail on an entity whose definition we don’t have.

Reply:
We bail out on them. To ignore them, we'd need to be able to parse them. I'll update the comment to specify this more precisely.

Reply:
I've updated the comment